File: 11_sieve_of_eratosthenes/count_semiprimes.py
def solution(n, p, q):
  prime_check = [0] * 2 + [1] * (n - 1)
  i = 2
  while i * i <= n:
    if prime_check[i] == 1:
      j = 2
      while i * j < n + 1:
        prime_check[i * j] = 0
        j += 1
    i += 1

  semiprime_check = [0] * (n + 1)
  i = 2
  while i < n + 1:
    if prime_check[i] == 0:
      i += 1
      continue
    j = i
    while j < n + 1:
      if prime_check[j] == 0:
        j += 1
        continue
      elif i * j > n:
        break
      else:
        semiprime_check[i * j] = 1
      j += 1
    i += 1

  prefix_count = [0] * (n + 1)
  i = 0
  cnt = 0
  while i < n + 1:
    cnt += semiprime_check[i]
    prefix_count[i] = cnt
    i += 1

  m = len(p)
  ret = []
  for i in xrange(m):
    ret.append(prefix_count[q[i]] - prefix_count[p[i]-1])
  return ret

File: 11_sieve_of_eratosthenes/count_semiprimes.cpp
#include <vector>

using namespace std;

vector<int> prime_check(int n) {
  vector<int> ret(n + 1, 1);
  ret[0] = 0;
  ret[1] = 0;
  for (int i = 2; i < n; i++) {
    if (ret[i] == 1) {
      for (int j = 2; i * j < n + 1; j++) {
        ret[i * j] = 0;
      }
    }
  }
  return ret;
}

vector<int> semiprime_cnts(int n) {
  vector<int> pcheck = prime_check(n);
  vector<int> ret(n + 1, 1);
  int cnt = 0;
  ret[0] = 0;
  ret[1] = 0;
  for (int i = 2; i < n; i++) {
    if (pcheck[i] == 1) {
      ret[i] = 0;
    }
    else if (ret[i] == 1) {
      for (int j = 2; i * j < n + 1; j++) {
        ret[i * j] = 0;
      }
    }
  }
  for (int i = 0; i < n + 1; i++) {
    if (ret[i] == 1) {
      cnt++;
    }
    ret[i] = cnt;
  }
  return ret;
}

vector<int> solution(int n, vector<int> &p, vector<int> &q) {
  int m = p.size();
  vector<int> ret(m, 0);
  vector<int> cnts = semiprime_cnts(n);
  for (int i = 0; i < m; i++) {
    ret[i] = cnts[q[i]] - cnts[p[i] - 1];
  }
  return ret;
}
